### PR TITLE
Azurefiles

### PR DESCRIPTION
**Pull request by moong-dal**
#### What is the purpose of this change?
Adding azurefiles as a backend

#### Was the change discussed in an issue or in the forum before?
https://github.com/rclone/rclone/issues/365#issuecomment-1755155134

#### Checklist

- [ Yes] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ Yes] I have added tests for all changes in this PR if appropriate.
- [ Yes/very sparse] I have added documentation for the changes if appropriate.
- [ No] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages). 
- [ Maybe] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by moong-dal** [comment]
Thanks for the prompt reply

### Do tests pass? 3/4
I had been running tests for `backend\azurefiles`, `fs/sync` and `fs/operations` which passed.
I had not run `test_all` earlier since I believed it was an alternative to individually running the three aforementioned sets of tests. Results from running `test_all -backends azurefiles -remotes 'TestAzureFiles:'` are attached. 
[2023-10-20-201815.zip](https://github.com/rclone/rclone/files/13055516/2023-10-20-201815.zip) The results are that tests named `backend\azurefiles`, `fs/sync` and `fs/operations` pass. However tests named 'vfs' fail. I have not attempted to implement any VFS functionality. 
Should implementing VFS be part of this pull request?

### Duplication of tests
Indeed there is duplication. I had written some tests of my own since I found it easier to reason through logic. I have deleted them. However, I have kept a small test for testing authentication with azure. This test is normally skipped since secret credentials are required to perform the test.

### Authenticating to Azure Files while using github CI
Without credentials authenticating and authorizing access to a file share on azure files, the integration tests will fail.
As maintainer of the repo, do you make a new remote in the config for each new backend?
Do you have credentials from Azure?

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (ncw) [inline .vscode/launch.json:1]: Please remove this!

File: README.md [hunk 2]
```diff
@@ -1,150 +1 @@
-[<img src="https://rclone.org/img/logo_on_light__horizontal_color.svg" width="50%" alt="rclone logo">](https://rclone.org/#gh-light-mode-only)
```
Comment (ncw) [inline README.md:1]: Can you fixup the README also please.

File: backend/azurefiles/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+*.exe
```
Comment (ncw) [inline backend/azurefiles/.gitignore:1]: Can we lose this? Or maybe put it in the main .gitignore

**Comment by ncw** [comment]
@moong-dal 

> ### Do tests pass? 3/4
> I had been running tests for `backend\azurefiles`, `fs/sync` and `fs/operations` which passed. I had not run `test_all` earlier since I believed it was an alternative to individually running the three aforementioned sets of tests. Results from running `test_all -backends azurefiles -remotes 'TestAzureFiles:'` are attached. [2023-10-20-201815.zip](https://github.com/rclone/rclone/files/13055516/2023-10-20-201815.zip) The results are that tests named `backend\azurefiles`, `fs/sync` and `fs/operations` pass. However tests named 'vfs' fail. I have not attempted to implement any VFS functionality. Should implementing VFS be part of this pull request?

You are doing extremely well :-) You only have one failing test out of the entire test suite!

You don't have to implement anything special for VFS - once your backend is correct the VFS tests will work too.

The error comes from the `TestFileReadAtZeroLength` test in particular 

https://github.com/rclone/rclone/blob/5b0d3d060fb269e7bb8e1d074237221edb250612/vfs/write_test.go#L380

which calls

https://github.com/rclone/rclone/blob/5b0d3d060fb269e7bb8e1d074237221edb250612/vfs/write_test.go#L367-L371

The problem here is that your backend is adding a Range request to a 0 length file. There is no valid Range request to request 0 bytes.

So in your Object.Open() method, don't add the  `file.HTTPRange{}` for a 0 length file and that will fix it I think.


> ### Duplication of tests
> Indeed there is duplication. I had written some tests of my own since I found it easier to reason through logic. I have deleted them. However, I have kept a small test for testing authentication with azure. This test is normally skipped since secret credentials are required to perform the test.

Sounds perfect :-)
 
> ### Authenticating to Azure Files while using github CI
> Without credentials authenticating and authorizing access to a file share on azure files, the integration tests will fail. 

On Github the integration tests won't be run - the `fstests.Run` will skip them.

These are run on the integration tester which runs daily - the output of which is here: https://pub.rclone.org/integration-tests/current/

> As maintainer of the repo, do you make a new remote in the config for each new backend?

Yes, on the integration test machine.

> Do you have credentials from Azure?

I have an azure account for the azureblob integration testing, I assume with enough poking in the azure console I can make some credentials for azurefiles!

So I think next steps are
- fix remaining test so `test_all` runs clean.
- get the CI running clean (no errors)
- rebase on master
- squash the commits if you like.

After that ping me here and I'll give the code another review.

There are still some docs missing - if you would like me to help with the docs I can do.

You are doing really well - we are nearly there :-)

**Comment by ncw** [comment]
I have pulled this code locally. I squashed and rebased it. The history was really weird so I ended up having to cherry-pick all the commits into a new branch.

I've got it running and it is working very well - well done :-)

I will take over now and
- fix the remaining test
- finish the documentation

I'll aim to merge for v1.65

Thank you!

**Comment by ncw** [comment]
Here is a test version of the completed backend.

[v1.65.0-beta.7491.b7119d10f.azurefiles](https://beta.rclone.org/branch/azurefiles/v1.65.0-beta.7491.b7119d10f.azurefiles/) on branch [azurefiles](https://github.com/rclone/rclone/tree/azurefiles) (uploaded in 15-30 mins)

This has these changes

    - use rclone's http Transport
    - fix handling of 0 length files
    - combine into one file and remove uneeded abstraction
    - make `chunk_size` and `upload_concurrency` settable
    - make auth the same as azureblob
    - set the Features correctly
    - implement `--azurefiles-max-stream-size`
    - remove arbitrary sleep on Mkdir
    - implement `--header-upload`
    - implement read and write MimeType for objects
    - implement optional methods
        - About
        - Copy
        - DirMove
        - Move
        - OpenWriterAt
        - PutStream
    - finish documentation

Thank you very much for your work .

I'll aim to merge this in the next day or two. If you find a bug, please send a pull request - thank you!

**Comment by ncw** [comment]
This is all merged now - thank you @moong-dal 

Please send followups as a new PR.

**Pull request closed without merging**